Foundation/NSURLSession/MultiHandle.swift

```diff
@@ -99,25 +103,33 @@ fileprivate extension URLSession._MultiHandle {
         // through libdispatch (DispatchSource) and store the source(s) inside
         // a `SocketSources` which we in turn store inside libcurl's multi handle
         // by means of curl_multi_assign() -- we retain the object fist.
-        let action = _SocketRegisterAction(rawValue: CFURLSessionPoll(value: what))
-        var socketSources = _SocketSources.from(socketSourcePtr: socketSourcePtr)
-        if socketSources == nil && action.needsSource {
-            let s = _SocketSources()
-            let p = Unmanaged.passRetained(s).toOpaque()
-            CFURLSessionMultiHandleAssign(rawHandle, socket, UnsafeMutableRawPointer(p))
-            socketSources = s
-        } else if socketSources != nil && action == .unregister {
-            // We need to release the stored pointer:
-            if let opaque = socketSourcePtr {
-                Unmanaged<_SocketSources>.fromOpaque(opaque).release()
+        commandQueue.async {
+            self.cancelInProgress?.wait()
```

Comment:
In my experience, I haven't seen `DispatchSource.cancel()` blocking for a long time. Moreover, the intention is to block the command queue until a `cancel()` goes through completely, of course assuming that `cancel()` will not block. I have considered the fact that the cancel handler runs on the same queue as the event handler, and also confirmed that the event handler will not block the cancel handler.

@dgrove-oss Can `DispatchSource.cancel()` can block under any other circumstances?